File: 2019W15/preppindataw15.R
library(dplyr)
library(readr)

allFiles <- list.files(path="E:/PD15/", full.names = TRUE, pattern = "Stock Purchases.csv$")
temp <- lapply(allFiles, read_csv, col_types =cols(`Customer ID` = col_double(),
                                                    `First Name` = col_character(),
                                                    `Last Name` = col_character(),
                                                    Sales = col_double(),
                                                    `Order Date` = col_character(),
                                                    Stock = col_character()))
names(temp) <- sub('.*PD15/(\\w+)\\s.*','\\1', allFiles)
for(i in names(temp)) 
  temp[[i]]$Region = i

final <-do.call(rbind, temp) %>%
  group_by(`Stock`,`Region`) %>%
  mutate('Total Regional Sales' = sum(`Sales`),
         '% of Total Regional Sales' = `Sales`/sum(`Sales`)*100) %>%
  group_by(`Stock`) %>%
  mutate('Total Sales' = sum(`Sales`),
         '% of Total Sales' = `Sales`/sum(`Sales`)*100) %>%
  filter(`% of Total Regional Sales`< 100) 

View(final)

#check <- final %>%
#  filter(`Region`=='South' & `Customer ID`=='1')

File: 2019/2019W15/PreppindataW15.py
import pandas as pd
import os
import re

os.chdir(r'E:/PD15/')
mergedata = []
for files in [f for f in os.listdir('.') if re.match('.*Stock Purchases.csv', f)]:
    dataset = pd.read_csv(files)
    dataset['Region']= re.search('^\w+(?=\s)',files).group(0)
    mergedata.append(dataset)    
df = pd.concat(mergedata)   

total = ((df.groupby(['Stock','Region']).agg({'Sales':'sum'}).rename(columns={'Sales': 'Total Regional Sales'}))
        .join(df.groupby(['Stock']).agg({'Sales':'sum'})).rename(columns={'Sales': 'Total Sales'})
        .reset_index()
        )

final = pd.merge(df, total, how='inner', on=['Region','Stock'])
final = final[final['Sales']<final['Total Regional Sales']]
final['% of Total Sales'] = final['Sales']/final['Total Sales']*100
final['% of Total Regional Sales'] = final['Sales']/final['Total Regional Sales']*100

#check = final.loc[(final['Region']=='South') & (final['Customer ID']==1)]
